consume.go

```diff
@@ -39,49 +39,130 @@ type consumeCmd struct {
 	poms          map[int32]sarama.PartitionOffsetManager
 }
 
-var offsetResume int64 = -3
+const (
+	maxOffset    int64 = 1<<63 - 1
+	offsetResume int64 = -3
+)
 
-type offset struct {
-	relative bool
-	start    int64
-	diff     int64
+type positionRange struct {
+	startAnchor, endAnchor anchor
+	diff                   anchorDiff
 }
 
-func (cmd *consumeCmd) resolveOffset(o offset, partition int32) (int64, error) {
-	if !o.relative {
-		return o.start, nil
+func (r positionRange) start() position {
+	return position{
+		anchor: r.startAnchor,
+		diff:   r.diff,
 	}
+}
 
-	var (
-		res int64
-		err error
-	)
+func (r positionRange) end() position {
+	return position{
+		anchor: r.endAnchor,
+		diff:   r.diff,
+	}
+}
+
+// position represents a position within the Kafka stream.
+// It holds an anchor (an absolute position specified as a
+// time stamp or an offset) and a relative position with
+// respect to that (specified as an offset delta or a time duration).
+type position struct {
+	anchor anchor
+	diff   anchorDiff
```

Comment:
Why not using an `anchor` instead of an `anchorDiff` here?

Reply:
Because a relative offset is specified differently from an absolute offset. It doesn't make sense to say `latest-[2019-07-05]` for example. Relative offsets can be specified as a signed numeric value or a time duration. Absolute offsets can be specified as an absolute unsigned numeric value, a timestamp, or a value like `latest`.